src/coap_gnutls.c

```diff
@@ -770,6 +774,98 @@ setup_pki_credentials(gnutls_certificate_credentials_t *pki_credentials,
     }
     break;
 
+  case COAP_PKI_KEY_PEM_BUF:
+    if (setup_data->pki_key.key.pem_buf.public_cert &&
+        setup_data->pki_key.key.pem_buf.public_cert_len &&
+        setup_data->pki_key.key.pem_buf.private_key &&
+        setup_data->pki_key.key.pem_buf.private_key_len) {
+      gnutls_datum_t cert;
+      gnutls_datum_t key;
+      int alloced_cert_memory = 0;
+      int alloced_key_memory = 0;
+
+      cert.size = setup_data->pki_key.key.pem_buf.public_cert_len;
+      if (setup_data->pki_key.key.pem_buf.public_cert[cert.size-1] != '\000') {
```

Comment:
Looking at the example code in `coap-server.c` for the new option `-m`, I do not see a guarantee that `cert.size > 0`.

Reply:
It is possible that the file being read into memory has a zero size, but coap-server:read_file_size() returns the file size + 1, unless the file is unreadable in which case it is 0.

several lines above in the code we have
````
    if (setup_data->pki_key.key.pem_buf.public_cert &&
        setup_data->pki_key.key.pem_buf.public_cert_len &&
````
so cert.size will never be 0. If setup_data->pki_key.key.pem_buf.public_cert_len is 0 and a server, this will get rejected on the else statement.  If it is a client, having no certs is valid.

Reply:
Ah, right. Sorry, I have missed the precondition for some reason.